src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java

```diff
@@ -0,0 +1,159 @@
+package fi.aalto.cs.apluscourses.intellij.utils;
+
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.intellij.model.IntelliJModuleMetadata;
+import fi.aalto.cs.apluscourses.model.Module;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.time.ZonedDateTime;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.io.FileUtils;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+public class CourseFileManager {
+
+  private File courseFile;
+  private URL courseUrl;
+  private Map<String, IntelliJModuleMetadata> modulesMetadata;
+
+  private CourseFileManager() {
+
+  }
+
+  private static final CourseFileManager instance = new CourseFileManager();
+
+  private static final String COURSE_FILE_NAME = "a-plus-project.json";
+
+  public static CourseFileManager getInstance() {
+    return instance;
+  }
+
+  /**
+   * Attempts to create a course file and load it for the given project with the given URL. If a
+   * course file for already exists (even with a different URL), this method loads the existing
+   * course file.
+   * @param project      The project for which the course file is created.
+   * @param courseUrl    The URL that gets added to the course file (if it doesn't exist yet).
+   * @throws IOException If an IO error occurs.
+   */
+  public synchronized void createAndLoad(@NotNull Project project, @NotNull URL courseUrl)
+      throws IOException {
+    courseFile = Paths
+        .get(project.getBasePath(), Project.DIRECTORY_STORE_FOLDER, COURSE_FILE_NAME)
+        .toFile();
+    if (!courseFile.exists()) {
+      this.courseUrl = courseUrl;
+      this.modulesMetadata = new HashMap<>();
+      JSONObject jsonObject = new JSONObject().put("url", courseUrl.toString());
+      FileUtils.writeStringToFile(courseFile, jsonObject.toString(), StandardCharsets.UTF_8);
+    } else {
+      internalLoad();
+    }
+  }
+
+  /**
+   * Attempts to load the course file corresponding to the given project. Returns {@code false} if
+   * the course file doesn't exist, {@code true} otherwise.
+   * @param project The project from which the course file is loaded.
+   * @return {@code true} if the course file was successfully loaded, {@code false} if the course
+   *         file doesn't exist.
+   * @throws IOException   If an IO error occurs while reading the course file.
+   * @throws JSONException If the course file contains malformed JSON.
+   */
+  public synchronized boolean load(@NotNull Project project) throws IOException {
+    courseFile = Paths
+        .get(project.getBasePath(), Project.DIRECTORY_STORE_FOLDER, COURSE_FILE_NAME)
+        .toFile();
```

Comment:
here is a minor code overlap with `createAndLoad(...)` (lines 49-51)

btw, would it make sense, that if the course file does not exist, then `createAndLoad(...)` will delegate to `load(...)`?

Reply:
Yep thats a good idea :+1: